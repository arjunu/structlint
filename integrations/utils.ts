import { exec } from "child_process";
import { resolve } from "path";

const structLint = ({ cwd }: { cwd: string }) =>
  new Promise((pResolve, pReject) => {
    exec(
      resolve(__dirname, "../bin/structlint.js"),
      {
        cwd,
        env: {
          ...process.env,
          // Turn off chalk color support
          // to make shapshot more readable
          FORCE_COLOR: "0",
        },
      },
      (error, stdout) => {
        if (!stdout) {
          pReject(new Error(`Failed to run structlint. Error: ${error}`));
        }

        pResolve({ stdout, code: error ? error.code : 0 });
      },
    );
  });

export { structLint };
