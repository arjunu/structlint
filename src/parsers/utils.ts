const getLanguageExtensions = (languages: string[]) => {
  return languages
    .map(lang => require(`linguist-languages/data/${lang}`))
    .flatMap(data => data.extensions);
};

export { getLanguageExtensions };
