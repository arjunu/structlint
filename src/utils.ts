import { partition } from "lodash/fp";
import { sep } from "path";
import * as globby from "globby";

/**
 *  Cast a `value` to exclude `null` and `undefined`.
 *  Throws if either `null` or `undefined` was passed
 */
function nonNullable<T>(
  value: T,
  message?: string,
): asserts value is NonNullable<T> {
  if (value === undefined || value === null) {
    throw new Error(
      message || `Non nullable values expected, received ${value}`,
    );
  }
}

type Primitive = string | number | boolean | undefined | null;

type DeepReadonly<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
  ? ReadonlyArray<U>
  : T extends Function
  ? T
  : DeepReadonlyObject<T>;

export type DeepReadonlyObject<T> = {
  readonly [P in keyof T]: DeepReadonly<T[P]>;
};

type TOptions = { expandDirectories: boolean };

const getPathFilesAndDirectories = async (
  path: string,
  { expandDirectories }: TOptions,
) => {
  const filesAndFolders = await globby(path, {
    expandDirectories,
    onlyFiles: false,
    gitignore: true,
    markDirectories: true,
  });

  const [directories, files] = partition(
    path => path.endsWith(sep),
    filesAndFolders,
  );

  return { directories, files };
};

export { nonNullable, Primitive, DeepReadonly, getPathFilesAndDirectories };
