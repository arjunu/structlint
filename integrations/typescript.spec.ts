import { resolve } from "path";

import { structLint } from "./utils";

describe("typescript support", () => {
  it("should return 2 violations", async () => {
    const stdout = await structLint({
      cwd: resolve(__dirname, "../examples/violations-ts"),
    });

    expect(stdout).toMatchSnapshot();
  });
});
