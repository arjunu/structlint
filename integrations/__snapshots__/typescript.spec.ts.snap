// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript support should return 2 violations 1`] = `
"Resolving configs...
Linting folder structure...
Found 4 violations

DISALLOWED IMPORTS: The following imports are disallowed in /modules/utils.ts
- react

DISALLOWED IMPORTS: The following imports are disallowed in /modules/transactions/sagas.ts
- /components/TxModal
- /components/TxSuccessModal
Business logic should never import UI components

DISALLOWED IMPORTS: The following imports are disallowed in /store.ts
- react

DISALLOWED IMPORTS: The following imports are disallowed in /components/shared/Button.ts
- /components/TxSuccessModal
"
`;
